### PR TITLE
[C++ API][Docs] Improve docs for torch::nn::Module

### PR DESCRIPTION
**Pull request by goldsborough**
Added some documentation. Will rebuild docs to make sure it looks good. Can already accept approvals.

@ebetica @apaszke @ezyang

### CODE REVIEW

**Review by soumith — APPROVED** [review]

File: torch/csrc/api/include/torch/nn/module.h [hunk 1]
```diff
@@ -37,30 +80,83 @@ class Module {
...
 
   /// Provides a means to traverse the `Module` tree.
+  ///
+  /// See the documentation for `CursorBase` for information on how to operate
+  /// on the returned cursor.
   ModuleCursor modules();
+
+  /// Provides a means to traverse the `Module` tree.
+  ///
+  /// See the documentation for `CursorBase` for information on how to operate
```
Comment (ezyang) [inline torch/csrc/api/include/torch/nn/module.h]: OOC: is it necessary to backtick class names? I was reading the Doxygen format manual, and it seemed to claim that ANY camelcase identifier would get hyperlinked if there was class of the same name.

File: torch/csrc/api/include/torch/nn/module.h [hunk 2]
```diff
@@ -108,31 +215,106 @@ class Module {
   void load(Archive& ar);
 
   /// Attempts to cast this `Module` to the given `ModuleType`.
+  ///
+  /// This method is useful when calling `apply()` on a `ModuleCursor`.
+  /// \rst
```
Comment (ezyang) [inline torch/csrc/api/include/torch/nn/module.h]: Other idle wondering: should we just turn on ReST by default?

**Review by ezyang — APPROVED** [review]

File: torch/csrc/api/include/torch/nn/module.h (hunk 1 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/module.h]: Things do get hyperlinked automatically, I just have a tendency to backtick any code things I mention. Maybe not necessary indeed

File: torch/csrc/api/include/torch/nn/module.h (hunk 2 above)
Comment (goldsborough) [reply torch/csrc/api/include/torch/nn/module.h]: ReSt is so cumbersome to read and write compared to Markdown IMO, just like how code has to be double backticked for example

**Pull request closed without merging**